### ISSUE TITLE
Speed improvement around matching **

### ISSUE THREAD

**Issue by jfmengels** [miniBill/elm-glob#2]
[Currently](https://github.com/miniBill/elm-glob/blob/330ee37e9cb43f3ba7b0a4d14a8a6a35993adc06/src/Glob.elm#L74-L75), when we encounter `**`, we call the matching function recursively, hope it works, and if it doesn't, ignore it.

```elm
matchComponents : List Component -> List String -> Bool
matchComponents components segments =
    case ( components, segments ) of
        -- ...other branches...

        ( TwoAsterisks :: ctail, _ :: stail ) ->
            if matchComponents components stail then
                True

            else
                matchComponents ctail segments
```

That seems inefficient to me, and I was thinking we could do this better.

Let's say we have the glob `**/c/d/bar.css` and the file path `a/b/c/d/bar.css`.
The idea is that (once we've reached the `**`) we could cut folders from the file path that we know we can ignore. In this case, we know the remaining part after `**/` is `c/d/bar.css`, so we can cut folders from the input path until we have 2 folder levels (for `c/` and `d/`).

Roughly, something like this:

```elm
        ( TwoAsterisks :: ctail, _ ) ->
            let
                numberOfExpectedLevels = List.length ctail
                
                numberOfLevelsRemaining = List.length segments
            in
            if numberOfExpectedLevels > numberOfLevelsRemaining then
                True

            else
                matchComponents ctail (List.drop (numberOfLevelsRemaining - numberOfExpectedLevels) segments)
```

This would only work if this is the last remaining `**` though. If it isn't, this optimization will not work.
Bonus points for computing whether this is the last `**` and maybe for precomputing the lengths of the remaining components and segments.

What do you think?

**Comment by miniBill** [comment]
Yeah, I think it breaks down for multiple `**`s. What is definitely safe is:
- pattern matching on `** :: **` and just dropping the first  `**`;
- `** :: []` and just matching if segments is nonempty (or maybe matching always, would need to think a bit more);
- `** :: Regex` and dropping segments until the regex matches
